### PR TITLE
RequirementCategory: implement `requirement_move` command 

### LINKED ISSUE #169: As a user, I want to move modules from a requirement category to another requirement category
so that I am easily able to move then around when I want to change their requirement category

### PR DESCRIPTION
**Pull request by chyeo**
Resolves #169.

Currently, our application does not allow users to move the modules that
have been added to requirement categories.

This makes it difficult for users when they are moving the modules in
the requirement category around. They would have to delete and add the
modules multiple times during the moving process.

To address these issues, let's add the 'requirement_move' command to
allow users to move modules in the requirement category.

First, let's implement the logic to handle how the modules are going to
be moved from one requirement category to another.

* [1/3] [Logic: create logic for 'requirement_move' command](https://github.com/CS2113-AY1819S2-T09-1/main/pull/171/commits/a7de2dd3afdf5b1d2b00852cabba44ab2613bb7e)
* [2/3] [Logic: update application to take in new command](https://github.com/CS2113-AY1819S2-T09-1/main/pull/171/commits/7778aacb6f6a85cda9a535c0d8c2e3261ea0f534)
* [3/3] [test/parser: creating test cases for new command](https://github.com/CS2113-AY1819S2-T09-1/main/pull/171/commits/434f419b5915d6603246d99c55b122f67533d438)

### CODE REVIEW

**Review by lycjackie — APPROVED** [review]
LGTM

I shall add @Creastery as reviewer for the model part.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+        this.toFind = name;
+        this.toMove.addAll(codeSet);
+    }
+
+    @Override
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
+        requireNonNull(model);
+
+        requirementCategories = model.getApplication().getRequirementCategoryList();
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Why is `requirementCategories` a class attribute?
You are not using the variable elsewhere, so you should limit the scope of the variable to this method only.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        RequirementCategory destinationRequirementCategory = model.getRequirementCategory(toFind);
+
+        if (destinationRequirementCategory == null) {
+            throw new CommandException(String.format(MESSAGE_NONEXISTENT_REQUIREMENT_CATEGORY, toFind));
+        }
+
+        if (toMove.stream().anyMatch(code -> !model.hasModuleCode(code))) {
+            String nonExistentCodeErrorMessage = toMove.stream().filter(code -> !model.hasModuleCode(code))
+                   .map(Code::toString).collect(Collectors.joining(", "));
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Here you are essentially repeating the same logic just to get the module codes.
Why not do it as such:
```java
Stream<Module> nonExistentCodes = toMove.stream().filter(code -> !model.hasModuleCode(code));
if (nonExistentCodes.count() > 0) {
    String nonExistentCodesErrorMessage = toMove.map(Code::toString).collect(Collectors.joining(", "));
    throw new CommandException(String.format(MESSAGE_NONEXISTENT_CODE, nonExistentCodeErrorMessage));
}
```

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            throw new CommandException(String.format(MESSAGE_NONEXISTENT_CODE, nonExistentCodeErrorMessage));
+        }
+
+        boolean codeInAnyRequirementCategory = toMove.stream()
+                .allMatch(code -> requirementCategories.stream()
+                .anyMatch(requirementCategory -> requirementCategory.getCodeSet().contains(code)));
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Indentation for nested lambda expressions is wrong.
Should have an additional indentation block at line 78.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        if (!codeInAnyRequirementCategory) {
+            String nonExistentCodeErrorMessage = toMove.stream().filter(code -> requirementCategories.stream()
+                    .anyMatch(requirementCategory -> !requirementCategory.getCodeSet().contains(code)))
+                    .map(Code::toString).collect(Collectors.joining(", "));
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java:107]: See previous comment on repetitive code logic and how you can simplify it.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+                    //reinitialize the updated destination requirement category
+                    destinationRequirementCategory = model.getRequirementCategory(toFind);
+                }
+            }
+        }
+
+        model.commitApplication();
+        return new CommandResult(String.format(MESSAGE_SUCCESS, toMove.stream().map(Code::toString).collect
+                (Collectors.joining(", ")), destinationRequirementCategory.getName()));
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Should move `.collect` to Line 127, or extract out the second argument to `String.format(...)` as a variable to remain consistent with the rest of your method.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                    .map(Code::toString).collect(Collectors.joining(", "));
+            throw new CommandException(
+                    String.format(MESSAGE_CODE_NOT_IN_ANY_REQUIREMENT_CATEGORY, nonExistentCodeErrorMessage));
+        }
+
+        RequirementCategory singleSourceRequirementCategory = requirementCategories.stream()
+                .filter(requirementCategory -> requirementCategory.getCodeSet()
+                .containsAll(toMove)).findFirst().orElse(null);
+
+        if (singleSourceRequirementCategory != null) {
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java:117]: What is the rationale for specifically handling the case of moving module codes from a single requirement category?
Slight performance boost?

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                //Check if destination code contains codes to move for edge cases
+                if (!destinationRequirementCategory.getCodeSet().contains(code)) {
+                    RequirementCategory sourceRequirementCategory = requirementCategories.stream()
+                            .filter(reqCat -> reqCat.getCodeSet().contains(code)).findFirst().orElse(null);
+
+                    RequirementCategory editedSourceRequirementCategory = RequirementCategoryUtil
+                            .getRequirementCategoryWithCodeRemoved(sourceRequirementCategory, code);
+                    RequirementCategory editedDestinationRequirementCategory =
+                            RequirementCategoryUtil
+                                    .getRequirementCategoryWithCodeAdded(destinationRequirementCategory, code);
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: This can be compressed to 2 lines only.

File: src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import pwe.planner.model.module.Code;
+import pwe.planner.model.requirement.RequirementCategory;
+
+/**
+ * A utility class for RequirementCategory.
+ */
+public class RequirementCategoryUtil {
+
+    /**
+     * Returns a RequirementCategory for the RequirementMoveCommand {@code RequirementMoveCommand}
```
Comment (Creastery) [inline src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java]: Using the same comment for all 4 methods in this class doesn't help to understand what the method is doing!

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+    private CommandHistory commandHistory = new CommandHistory();
+    private Model model;
+
+    @Before
+    public void setUp() throws IllegalValueException {
+        model = new ModelManager(
+                new JsonSerializableApplication(getTypicalModuleList(), getTypicalDegreePlannerList(),
+                        getTypicalRequirementCategoriesList()).toModelType(),
+                new UserPrefs());
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: Compress this to 3 lines.

**Review by Creastery — CHANGES_REQUESTED** [review]

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java (hunk 9 above)
Comment (chyeo) [reply src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: Have compressed the statments to 3 lines.

File: src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java (hunk 8 above)
Comment (chyeo) [reply src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java]: The comments for the methods have been changed.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 7 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Have compressed the statments

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 6 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java:117]: Yes. as we are moving from a single source to destination.
The alternative method would be to loop through all the codes specified by the user.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 5 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Have extracted out the `collector` to the `RequirementCategoryUtil`

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 1 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Have moved the variable down to the method

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 3 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Have removed this method and replaced it with another piece of code

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 2 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Roger, will simplify the affected codes

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 4 above)
Comment (chyeo) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java:107]: Roger, the codes have been simplified

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                    model.setRequirementCategory(sourceRequirementCategory, editedSourceRequirementCategory);
+                    model.setRequirementCategory(destinationRequirementCategory, editedDestinationRequirementCategory);
+
+                    //reinitialize the updated destination requirement category
+                    destinationRequirementCategory = model.getRequirementCategory(toFind);
+                }
+            }
+        }
+
+        String formattedCodeString = toMove.stream().map(Code::toString).collect(Collectors.joining(", "));
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: `formattedCodeString` -> `codesMoved`

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    public static final String MESSAGE_SUCCESS = "Successfully moved module(s) (%1$s) to requirement category %2$s!";
+
+    // Command failure messages
+    public static final String MESSAGE_NONEXISTENT_REQUIREMENT_CATEGORY =
+            "You cannot specify a requirement category (%1$s) that does not exist!\n"
+            + "Perhaps you misspelled the name of the requirement category?";
+
+    public static final String MESSAGE_CODE_NOT_IN_ANY_REQUIREMENT_CATEGORY =
+            "You cannot move module(s) (%1$s) that does not exist in any requirement category!\n"
+            + "Perhaps you were trying to add modules into the requirement category?";
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: The correct command `requirement_add` should be mentioned here to inform the user what command to use to add modules to a requirement category!.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+    public static final String MESSAGE_NONEXISTENT_CODE = "You cannot specify module(s) (%1$s) that does not exist!\n"
+            + "Perhaps you were trying to add modules into the application?\n"
+            + "[Tip] You can try adding the module into the application first.\n"
+            + "Afterwards, add the module into a requirement category.\n"
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Same as above.

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+            + "Then you will be able to move the module around the requirement categories!";
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: This line is not necessary since you are providing a tip that assumes that the module does not exists in the application.

File: src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import pwe.planner.model.module.Code;
+import pwe.planner.model.requirement.RequirementCategory;
+
+/**
+ * A utility class for RequirementCategory.
+ */
+public class RequirementCategoryUtil {
+
+    /**
+     * Returns a requirement category object with the specified code {@code toMove} removed from the
```
Comment (Creastery) [inline src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java]: Returns a **new** requirement category...

File: src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        Set<Code> newCodeSet = new HashSet<>(reqCat.getCodeSet());
+        newCodeSet.remove(toMove);
+        RequirementCategory requirementCategory = new RequirementCategory(reqCat.getName(), reqCat.getCredits(),
+                newCodeSet);
+
+        return requirementCategory;
+    }
+
+    /**
+     * Returns a requirement category object with the specified code set {@code toMove} removed from the
```
Comment (Creastery) [inline src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java]: Same as above.

File: src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        Set<Code> newCodeSet = new HashSet<>(reqCat.getCodeSet());
+        newCodeSet.removeAll(setToMove);
+        RequirementCategory requirementCategory = new RequirementCategory(reqCat.getName(), reqCat.getCredits(),
+                newCodeSet);
+
+        return requirementCategory;
+    }
+
+    /**
+     * Returns a requirement category object with the specified code {@code toMove} added from the
```
Comment (Creastery) [inline src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java]: Same as above.

File: src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        Set<Code> newCodeSet = new HashSet<>(reqCat.getCodeSet());
+        newCodeSet.add(toMove);
+        RequirementCategory requirementCategory = new RequirementCategory(reqCat.getName(), reqCat.getCredits(),
+                newCodeSet);
+
+        return requirementCategory;
+    }
+
+    /**
+     * Returns a requirement category object with the specified code set {@code toMove} added from the
```
Comment (Creastery) [inline src/main/java/pwe/planner/model/util/RequirementCategoryUtil.java]: Same as above.

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+
+    @Test
+    public void constructor_nullInputs_throwsNullPointerException() {
+        thrown.expect(NullPointerException.class);
+        new RequirementMoveCommand(null, null);
+    }
+
+    @Test
+    public void execute_nonExistentRequirementCategory_throwsCommandException() {
+        Set<Code> validCodeSet = new HashSet<>(Set.of(new Code("CS1231")));
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: There's no need to wrap `Set.of(...)` within `new HashSet<>(...)`!
Same goes for the remaining of this file.

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+        Name nonExistentRequirementCategoryName = new Name("DOES NOT EXIST");
+        assertCommandFailure(new RequirementMoveCommand(nonExistentRequirementCategoryName, validCodeSet),
+                model, commandHistory, String.format(RequirementMoveCommand.MESSAGE_NONEXISTENT_REQUIREMENT_CATEGORY,
+                        nonExistentRequirementCategoryName));
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: You can consider intialise the `command` and `expectedMessage` separately to keep the `assertCommandXYZ(...)` short.
Same goes for the remaining of this file.

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+                String.format(RequirementMoveCommand.MESSAGE_SUCCESS, formattedCodeString,
+                        destinationReqCatName), expectedModel);
+    }
+
+    @Test
+    public void execute_moveCodesFromMultipleSource_success() {
+        Set<Code> validCodeSet = new HashSet<>(Set.of(new Code("CS1231"), new Code("CS2100")));
+
+        Set<Code> sourceCodeSet1 = new HashSet<>();
+        sourceCodeSet1.add(new Code("CS2100"));
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: Could have just used `Set.of(...)` here.

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 21]
```diff
@@ -0,0 +1,285 @@
...
+
+        Set<Code> sourceCodeSet2 = new HashSet<>();
+        sourceCodeSet2.add(new Code("CS1231"));
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: Same as above.

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+
+    @Test
+    public void execute_moveCodesWithSameSourceAndDestination_success() {
+        Set<Code> validCodeSet = new HashSet<>(Set.of(new Code("CS1231"), new Code("CS2100")));
+
+        Set<Code> sourceCodeSet1 = new HashSet<>();
+        sourceCodeSet1.add(new Code("CS2100"));
+
+        Set<Code> sourceCodeSet2 = new HashSet<>();
+        sourceCodeSet2.add(new Code("CS1231"));
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: Same as above.

File: src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java [hunk 23]
```diff
@@ -0,0 +1,285 @@
...
+                        destinationReqCatName), expectedModel);
+    }
+
+    @Test
+    public void execute_equals() {
+        Set<Code> validCodeSet = new HashSet<>(Set.of(new Code("CS2100")));
+        Name requirementCategoryName = new Name("Computing Foundation");
+
+        Set<Code> codeSetToCompare = new HashSet<>();
+        codeSetToCompare.add(new Code("CS1010"));
```
Comment (Creastery) [inline src/test/java/pwe/planner/logic/commands/RequirementMoveCommandTest.java]: Same as above.

**Review by Creastery — CHANGES_REQUESTED** [review]

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+    // Command failure messages
+    public static final String MESSAGE_NONEXISTENT_REQUIREMENT_CATEGORY =
+            "You cannot specify a requirement category (%1$s) that does not exist!\n"
+            + "Perhaps you misspelled the name of the requirement category?";
+
+    public static final String MESSAGE_CODE_NOT_IN_ANY_REQUIREMENT_CATEGORY =
+            "You cannot move module(s) (%1$s) that does not exist in any requirement category!\n"
+            + "Perhaps you were trying to add modules into the requirement category?\n"
+            + "You can try adding the module into the requirement category using the 'requirement_add' command.";
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Add a `[Tip]` in front of this sentence!

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+
+    public static final String MESSAGE_NONEXISTENT_CODE = "You cannot specify module(s) (%1$s) that does not exist!\n"
+            + "Perhaps you were trying to add modules into the application?\n"
+            + "[Tip] You can try adding the module into the application first using the 'add' command.\n"
+            + "Afterwards, add the module into a requirement category.\n";
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Can rephrase this to:
```
[Tip] You may want to add the module into the module list first using the "add" command.
Afterwards, you can add the module into the requirement category using the "requirement_add" command!
```

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java (hunk 24 above)
Comment (Creastery) [reply src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: `'requirement_add'` -> `"requirement_add"`

**Review by Creastery — CHANGES_REQUESTED** [review]

File: src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+
+    // Command success message
+    public static final String MESSAGE_SUCCESS = "Successfully moved module(s) (%1$s) to requirement category %2$s!";
+
+    // Command failure messages
+    public static final String MESSAGE_NONEXISTENT_REQUIREMENT_CATEGORY =
+            "You cannot specify a requirement category (%1$s) that does not exist!\n"
+            + "Perhaps you misspelled the name of the requirement category?\n"
+            + "[Tip] You may want to refer to the requirement category list to see the requirement categories"
+            + " in the application!";
```
Comment (Creastery) [inline src/main/java/pwe/planner/logic/commands/RequirementMoveCommand.java]: Where can the user find this `requirement category list`?
I'd suggest to print the list of acceptable requirement category names here regardless.

**Review by Creastery — CHANGES_REQUESTED** [review]

**Review by Creastery — APPROVED** [review]
LGTM. Approving and merging.

**Comment by chyeo** [comment]
Thanks for the review!
